DuckDBRowReader: Add class to stream DuckDB results efficiently

rstefanic opened the pull request:
There currently exists the `DuckDBResultReader`, but the issue that we're running into is that this processes chunks into an internal buffer that's never cleared. For large result sets, this buffer builds and builds until we run out of memory.

The case that we have is that we just need the result set row by row without needing to compute any of the column data. This new class does less than `DuckDBResultReader` by not storing the results of the query in the underlying buffer. Instead, the results of the chunk are passed back to the caller where the caller can handle the chunks as they wish.

The interface for processing rows is similar to `DuckDBRowReader` where the caller can still use converters as they wish to process their data.

From a `DuckDBConnection`, calling `rowReaderStream` will return a `DuckDBRowReader` where the caller can handle each individually processed chunk without having it written to an internal buffer for more memory efficient operations. 

```js
const rowReader = await connection.rowReaderStream(sql);

for await (const rows of rowReader.getRowObjects()) {

  // `rows` is a Record<string, DuckDBValue>[] from the last chunk that was processed.
  // Here, we can console.log each row without keeping the entire result set in memory.
  // We process the chunk of `rows` which is only available in this scope.

  for (const row or rows) {
    console.log(row);
  }
}
```

I didn't see any documentation on contributing, but if this is a feature you're interested in supporting, I'd be happy to change the formatting, add tests, change the API of `DuckDBRowReader`, etc. since this is an issue that we're facing at my job.

jraymakers commented on api/src/DuckDBConnection.ts:
```diff
@@ -87,6 +88,13 @@ export class DuckDBConnection {
+    values?: DuckDBValue[] | Record<string, DuckDBValue>,
+    types?: DuckDBType[] | Record<string, DuckDBType | undefined>
+  ): Promise<DuckDBRowReader> {
+    return new DuckDBRowReader(await this.run(sql, values, types));
```
The `run` family of methods cause DuckDB to fully materialize the result in memory, so, for this purpose, it would be better to use `stream` below.

rstefanic replied:
Ah, I didn't know! That's really helpful.

jraymakers commented on api/src/DuckDBRowReader.ts:
```diff
@@ -0,0 +1,182 @@
+   * this class are wrappers around this generator so that the caller can specify
+   * how they want to consume the rows.
+   */
+  private async *getRows(): AsyncGenerator<DuckDBValue[][], void, undefined> {
```
This should be public so callers can get the raw (unconverted) value arrays if they want.

Additionally, we should expose a `getChunks` method that yields the raw chunks.

jraymakers commented on api/src/DuckDBRowReader.ts:
```diff
@@ -0,0 +1,182 @@
+/** Used to read the results row by row in a memory efficient manner */
+export class DuckDBRowReader {
+  private readonly result: DuckDBResult;
+  private chunks: AsyncGenerator<DuckDBDataChunk, void, undefined>;
```
This reference can be `readonly`, I believe.

jraymakers commented on api/src/DuckDBRowReader.ts:
```diff
@@ -0,0 +1,182 @@
+    }
+  }
+
+  public async *convertNextRowObjects<T>(
```
Naming nit: call this `convertRowObjects`. Using the term "next" isn't consistent with this being a generator or how the other methods are named.

jraymakers commented on api/src/DuckDBRowReader.ts:
```diff
@@ -0,0 +1,182 @@
+    undefined
+  > {
+    for await (const chunk of this.chunks) {
+      yield getRowObjectsFromChunks([chunk], this.deduplicatedColumnNames());
```
It would be slightly better to call `deduplicatedColumnNames` once before the loop, both here and in the next method below.

rstefanic replied:
Good call! I've since removed the DuckDBRowReader, but I've carried this suggestion over to the new methods added to the `DuckDBResult` class.

jraymakers reviewed (commented):
Thanks for the PR! I definitely support adding something like this to the API.

Another contributor submitted [this related PR](https://github.com/duckdb/duckdb-node-neo/pull/234) some time ago, but I haven't yet found the time to integrate it properly into the API by making it more general.

This PR comes closer, but I think we can do even better by combining the two approaches. I'm wondering if we can implement the async iterator protocol directly on `DuckDBResult`, as done in the above PR, but have it just yield chunks, instead of performing a particular conversion and yielding rows (as the above PR does).

Then, we can use this iterator to add other convenience methods to `DuckDBResult` that perform the various conversions and also return results row-by-row in addition to chunk-by-chunk. I think that would alleviate the need for a separate `DuckDBRowReader` class or a new method on `DuckDBConnection`.

Below are a few detailed comments on this PR, which may or may not be relevant if we adopt the alternate approach described above.

jraymakers commented:
Here are some examples of the API I'm envisioning.

Yield chunks:
```ts
const result = await connection.stream(sql);
for await (const chunk of result) {
  // ...
}
```

Yield rows (:
```ts
const result = await connection.stream(sql);
for await (const row of result.yieldRows()) {
  // ...
}
```

Yield row objects (:
```ts
const result = await connection.stream(sql);
for await (const rowObject of result.yieldRowObjects()) {
  // ...
}
```

With (general) conversion (:
```ts
const result = await connection.stream(sql);
for await (const row of result.yieldRows(converter)) {
  // ...
}
```

With specific conversion (:
```ts
const result = await connection.stream(sql);
for await (const row of result.yieldRowsJS()) {
  // ...
}
```

Plus other combinations for converting row objects, returning JSON, etc.

jraymakers commented on api/src/DuckDBResult.ts:
```diff
@@ -239,4 +239,56 @@ export class DuckDBResult {
+    }
+  }
+
+  public async *yieldConvertedRows<T>(
```
I think we want both `yieldConvertedRowObjects` and `yieldConvertedRows`. The implementation of this method looks like the former, because it calls `convertRowObjectsFromChunk`; the latter should call `convertRowsFromChunks`.

rstefanic replied:
Makes sense!

jraymakers commented on api/src/DuckDBResult.ts:
```diff
@@ -239,4 +239,56 @@ export class DuckDBResult {
+  }
+
+  public async *yieldRowsJs(): AsyncIterableIterator<Record<string, JS>[]> {
+    return this.convertRowObjects(JSDuckDBValueConverter);
```
I'd expect both `yieldRowsJs` and `yieldRowsJson` to call `yieldConvertedRows`.

Also, I think we want both pairs `yieldRowsJs`/`yieldRowsJson` and `yieldRowObjectsJs`/`yieldRowObjectJson`. The former should call `yieldConvertedRows`, while the latter should call `yieldConvertedRowObjects`.

rstefanic replied:
Ah! I get it now. I knew I was missing something. The `yieldRows*` methods convert the rows into JS/JSON while the `yieldRowObject*` methods convert them into objects.

Seems obvious now that I typed it out. Thanks!

jraymakers commented on api/src/DuckDBResult.ts:
```diff
@@ -239,4 +239,56 @@ export class DuckDBResult {
+
+  public async *yieldRows(): AsyncIterableIterator<DuckDBValue[][]> {
+    const iterator = this[Symbol.asyncIterator]();
+    for await (const chunk of iterator) {
```
Does `for await (const chunk of this)` work, instead of explicitly getting the async iterator?

rstefanic replied:
Didn't think of that. Clever 👍

jraymakers reviewed (commented):
I definitely like where this is headed; it's looking very clean. Just a few comments about the details and the set of variants.

jraymakers commented:
Also: Once we have all the methods in place, it would be good to add some unit tests for each to `api.test.ts`.

(I apologize for the single large unit test file. I've been meaning to split it up, but haven't gotten to it yet.)

rstefanic commented:
One thing to note that I forgot to add to my commit message is that I had to remove the `async` call before the JS/Json converter functions. It was returning a promise which was unnecessary and confusing.

I have added tests here, and they all passed on my branch when I ran them on my local machine; however there is now a conflict since my branch was a bit stale. On my machine, I have rebased my branch onto main and resolved the conflict, but the tests don't pass.

Even with a fresh pull of the repo from main has many of the tests failing. If you just want me to push my version with the merge conflict resolved, I can. But I wasn't sure if I wanted to push it since it was failing anyways.

jraymakers commented:
It seems the tests pass in this PR. Perhaps something is stale in your local environment? Have you tried `pnpm run clean` in both `bindings` and `api`?

jraymakers commented on api/test/api.test.ts:
```diff
@@ -2445,4 +2445,71 @@ ORDER BY name
+  test("iterate over DuckDBResult stream in chunks", async () => {
+    await withConnection(async (connection) => {
+      const result = await connection.stream(
+        "select i::int, i::int + 10, (i + 100)::varchar from range(3) t(i)",
```
It would be good to have a test that generates more than one chunk, in addition to this basic one that generates a single chunk. A version of this test that uses `range(5000)` or something would suffice.

jraymakers commented on api/test/api.test.ts:
```diff
@@ -2445,4 +2445,71 @@ ORDER BY name
     });
   });
+
+  test("iterate over DuckDBResult stream in chunks", async () => {
```
Would be good to have test cases for `yieldRows` and `yieldRowObjects`.

rstefanic replied:
I figured that the converters for JS and JSON covered these since they both call `yieldRows` and `yieldRowObjects`.

jraymakers reviewed (commented):
Looks great! Just a couple comments about some additional test cases.

jraymakers commented:
BTW I was able to pull your fork and run both the build and tests just fine, both on `main` and your branch.

rstefanic commented:
Your suggestion to run `pnpm run clean` and rebuild the bindings is what helped. I built those over a week ago which explains why my tests were failing. Thanks!

I added three new tests. One that generates more than one chunk, a test for `yieldRows`, and a test for `yieldRowObjects.
